File: 2025_05/Pro_388353.java
import java.util.*;

class Solution {
    static class Node {
        int x, y;
        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public int solution(String[] storage, String[] requests) {
        int n = storage.length;
        int m = storage[0].length();
        int answer = 0;

        char[][] map = new char[n][m];
        for (int i = 0; i < n; i++) {
            map[i] = storage[i].toCharArray();
        }

        for (String req : requests) {
            //지게차
            if (req.length() == 1) {
                char ch = req.charAt(0);
                boolean[][] visited = new boolean[n][m];
                boolean[][] access = new boolean[n][m];
                Queue<Node> q = new LinkedList<>();

                // BFS 시작점
                for (int i = 0; i < m; i++) {
                    if (map[0][i] == '\0'&&!visited[0][i]) {
                        q.offer(new Node(0, i));
                        visited[0][i] = true;
                    }
                    if (map[n - 1][i] == '\0'&&!visited[n-1][i]) {
                        q.offer(new Node(n - 1, i));
                        visited[n - 1][i] = true;
                    }
                }
                for (int i = 0; i < n; i++) {
                    if (map[i][0] == '\0'&&!visited[i][0]) {
                        q.offer(new Node(i, 0));
                        visited[i][0] = true;
                    }
                    if (map[i][m - 1] == '\0'&&!visited[i][m-1]) {
                        q.offer(new Node(i, m - 1));
                        visited[i][m - 1] = true;
                    }
                }

                // BFS 탐색: 빈칸 통해 접근 가능한 ch 체크
                while (!q.isEmpty()) {
                    Node node = q.poll();
                    int x = node.x;
                    int y = node.y;
                    access[x][y]=true;// 외부랑 접근가능한지 bfs를 통해 마킹한 배열
                    for (int d = 0; d < 4; d++) {
                        int nx = x + dx[d];
                        int ny = y + dy[d];
                        if (nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny]) {
                            if (map[nx][ny] == '\0') {
                                visited[nx][ny] = true;
                                q.offer(new Node(nx, ny));
                            }
                        }
                    }
                }
                
                

                // 외부와 연결된 ch인지 판단
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < m; j++) {
                        boolean acc=false;
                        if (map[i][j] == ch) {
                            if(i==0|| j==0||i==n-1||j==m-1){//가장자리니까 당연히 접근가능
                                acc=true;
                            }else{
                                for(int k=0;k<4;k++){
                                    int nx = i+dx[k];
                                    int ny = j+dy[k];
                                    if(nx>=0&&nx<n&&ny>=0&&ny<m){
                                        if(access[nx][ny]){
                                            acc=true;
                                        }
                                    }
                                }
                            }
                        }
                        if(acc){
                            map[i][j]='\0';
                        }
                    }
                }

            } else {
                // 크레인
                char ch = req.charAt(0);
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < m; j++) {
                        if (map[i][j] == ch) {
                            map[i][j] = '\0';
                        }
                    }
                }
            }
        }

        // 남은 문자 수 세기
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print(map[i][j]+" ");
                if (map[i][j] != '\0') answer++;
            }
            System.out.println();
        }

        return answer;
    }
}

File: 2025/2025_03/Pro_388353.cpp
#include <bits/stdc++.h>
using namespace std;
char arr[51][51];
int row, col;
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

int solution(vector<string> storage, vector<string> requests)
{
    row = storage.size();
    col = storage[0].size();
    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            arr[i][j] = storage[i][j];
        }
    }

    for (string req : requests)
    {
        if (req.size() == 1)
        {

            char target = req[0];
            bool visited[51][51] = {
                false,
            };
            bool accessible[51][51] = {
                false,
            };
            queue<pair<int, int>> q;
            // 빈칸들 먼저 연결하고, 접근 가능한지 판단
            for (int i = 0; i < col; i++)
            {
                if (!visited[0][i] && arr[0][i] == '\0')
                {
                    visited[0][i] = true;
                    q.push({0, i});
                }
                if (!visited[row - 1][i] && arr[row - 1][i] == '\0')
                {
                    visited[row - 1][i] = true;
                    q.push({row - 1, i});
                }
            }

            for (int i = 0; i < row; i++)
            {
                if (!visited[i][0] && arr[i][0] == '\0')
                {
                    visited[i][0] = true;
                    q.push({i, 0});
                }
                if (!visited[i][col - 1] && arr[i][col - 1] == '\0')
                {
                    visited[i][col - 1] = true;
                    q.push({i, col - 1});
                }
            }
            while (!q.empty())
            {
                int x = q.front().first;
                int y = q.front().second;
                q.pop();
                accessible[x][y] = true;
                for (int k = 0; k < 4; k++)
                {
                    int nx = x + dx[k];
                    int ny = y + dy[k];
                    if (nx >= 0 && nx < row && ny >= 0 && ny < col)
                    {
                        if (!visited[nx][ny] && arr[nx][ny] == '\0')
                        {
                            visited[nx][ny] = true;
                            q.push({nx, ny});
                        }
                    }
                }
            }
            // target인 지점 중에서 빈칸 연결인지 찾기
            for (int i = 0; i < row; i++)
            {
                for (int j = 0; j < col; j++)
                {
                    if (arr[i][j] == target)
                    {
                        bool acc = false;
                        if (i == 0 || j == 0 || i == row - 1 || j == col - 1)
                        {
                            acc = true;
                        }
                        else
                        {
                            for (int k = 0; k < 4; k++)
                            {
                                int nx = i + dx[k];
                                int ny = j + dy[k];
                                // 외부와 연결됨 중요!!
                                if (nx < 0 || nx >= row || ny < 0 || ny >= col)
                                {
                                    acc = true;
                                    break;
                                }
                                if (accessible[nx][ny])
                                {
                                    acc = true;
                                    break;
                                }
                            }
                        }
                        if (acc)
                        {
                            arr[i][j] = '\0';
                        }
                    }
                }
            }
        }
        else if (req.size() > 1)
        {
            char target = req[0];
            for (int i = 0; i < row; i++)
            {
                for (int j = 0; j < col; j++)
                {
                    if (arr[i][j] == target)
                    {
                        arr[i][j] = '\0';
                    }
                }
            }
        }
    }

    int answer = 0;
    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            if (arr[i][j] != '\0')
            {
                answer++;
            }
        }
    }
    return answer;
}
